services/search/pkg/engine/bleve.go

```diff
@@ -0,0 +1,366 @@
+			match.Entity.LastModifiedTime = &timestamppb.Timestamp{Seconds: mtime.Unix(), Nanos: int32(mtime.Nanosecond())}
+		}
+
+		matches = append(matches, match)
+	}
+
+	return &searchService.SearchIndexResponse{
+		Matches:      matches,
+		TotalMatches: int32(res.Total),
+	}, nil
+}
+
+// Upsert indexes or stores Resource data fields.
+func (b *Bleve) Upsert(id string, r Resource) error {
+	return b.index.Index(id, r)
+}
+
+// Move updates the resource location and all of its necessary fields.
+func (b *Bleve) Move(id string, parentid string, target string) error {
+	r, err := b.getResource(id)
+	if err != nil {
+		return err
+	}
+	oldName := r.Path
+	newName := utils.MakeRelativePath(target)
+
+	r, err = b.updateEntity(id, func(r *Resource) {
+		r.Path = newName
+		r.Name = path.Base(newName)
+		r.ParentID = parentid
+	})
+	if err != nil {
+		return err
+	}
+
+	if r.Type == uint64(storageProvider.ResourceType_RESOURCE_TYPE_CONTAINER) {
+		q := bleve.NewConjunctionQuery(
+			bleve.NewQueryStringQuery("RootID:"+r.RootID),
+			bleve.NewQueryStringQuery("Path:"+escapeQuery(oldName+"/*")),
+		)
+		bleveReq := bleve.NewSearchRequest(q)
+		bleveReq.Size = math.MaxInt
+		bleveReq.Fields = []string{"*"}
+		res, err := b.index.Search(bleveReq)
+		if err != nil {
+			return err
+		}
+
+		for _, h := range res.Hits {
+			_, err := b.updateEntity(h.ID, func(r *Resource) {
+				r.Path = strings.Replace(r.Path, oldName, newName, 1)
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Delete marks the resource as deleted.
+// The resource object will stay in the bleve index,
+// instead of removing the resource it just marks it as deleted!
+// can be undone
+func (b *Bleve) Delete(id string) error {
+	return b.setDeleted(id, true)
+}
+
+// Restore is the counterpart to Delete.
+// It restores the resource which makes it available again.
+func (b *Bleve) Restore(id string) error {
+	return b.setDeleted(id, false)
+}
+
+// Purge removes a resource from the index, irreversible operation.
+func (b *Bleve) Purge(id string) error {
+	return b.index.Delete(id)
+}
+
+// DocCount returns the number of resources in the index.
+func (b *Bleve) DocCount() (uint64, error) {
+	return b.index.DocCount()
+}
+
+func (b *Bleve) getResource(id string) (*Resource, error) {
+	req := bleve.NewSearchRequest(bleve.NewDocIDQuery([]string{id}))
+	req.Fields = []string{"*"}
+	res, err := b.index.Search(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.Hits.Len() == 0 {
+		return nil, errors.New("entity not found")
+	}
+
+	fields := res.Hits[0].Fields
+
+	return &Resource{
+		ID:       getValue[string](fields, "ID"),
+		RootID:   getValue[string](fields, "RootID"),
+		Path:     getValue[string](fields, "Path"),
+		ParentID: getValue[string](fields, "ParentID"),
+		Type:     uint64(getValue[float64](fields, "Type")),
+		Deleted:  getValue[bool](fields, "Deleted"),
+		Document: content.Document{
+			Name:     getValue[string](fields, "Name"),
```

Comment:
Shouldn't the `Document` also get the `Title` and the `Content` fields? Otherwise I would expect those to get lost when modifying entities in `updateEntity`, for example, no?

Reply:
good point, you are right. THX